src/software/ai_vs_ai_ci.py

```diff
@@ -0,0 +1,115 @@
+from software.thunderscope.proto_unix_io import ProtoUnixIO
+from software.thunderscope.thread_safe_buffer import ThreadSafeBuffer
+from software.thunderscope.binary_context_managers import *
+from software.thunderscope.replay.proto_logger import ProtoLogger
+from proto.message_translation import tbots_protobuf
+import software.python_bindings as cpp_bindings
+
+import numpy
+
+
+def start_ai_vs_ai(simulator_runtime_dir, blue_fs_dir, yellow_fs_dir):
+    """
+    Start AI vs AI silently, ticking as fast as possible without visualization.
+
+    Note: This module is different from Thunderscope as it does not run any Qt elements at all. It runs two
+    FullSystems, a Gamecontroller and TigersAutoref. Qt does not work with display-less devices, so it won't work
+    properly in a CI setting, which is why this AI vs AI setup is required.
+
+    :param simulator_runtime_dir    the runtime directory to set up ProtoUnixIO for the Simulator
+    :param blue_fs_dir              the runtime directory to set up ProtoUnixIO for the blue FullSystem
+    :param yellow_fs_dir            the runtime directory to set up ProtoUnixIO for the yellow FullSystem
+    """
+
+    blue_fs_proto_unix_io = ProtoUnixIO()
+    yellow_fs_proto_unix_io = ProtoUnixIO()
+    simulator_proto_unix_io = ProtoUnixIO()
+
+    def __async_sim_ticker():
+        """Setup the world and tick simulation forever, as fast as possible
+
+        :param tick_rate_ms: The tick rate of the simulation
+
+        """
+        yellow_world_buffer = ThreadSafeBuffer(buffer_size=1, protobuf_type=World)
+        blue_world_buffer = ThreadSafeBuffer(buffer_size=1, protobuf_type=World)
+
+        yellow_fs_proto_unix_io.register_observer(World, yellow_world_buffer)
+        blue_fs_proto_unix_io.register_observer(World, blue_world_buffer)
+
+        world_state_received_buffer = ThreadSafeBuffer(1, WorldStateReceivedTrigger)
+        simulator_proto_unix_io.register_observer(
+            WorldStateReceivedTrigger, world_state_received_buffer
+        )
+
+        while True:
+            world_state_received = world_state_received_buffer.get(
+                block=False, return_cached=False
+            )
+            if not world_state_received:
+                world_state = tbots_protobuf.create_default_world_state(
+                    DIV_B_NUM_ROBOTS
+                )
+                simulator_proto_unix_io.send_proto(WorldState, world_state)
+            else:
+                break
+
+            time.sleep(0.01)
+
+            while True:
+                # Tick Simulation
+                tick = SimulatorTick(
+                    milliseconds=DEFAULT_SIMULATOR_TICK_RATE_MILLISECONDS_PER_TICK
+                )
+                simulator_proto_unix_io.send_proto(SimulatorTick, tick)
+
+                # Wait for full systems to process worlds before continuing
+                yellow_world_buffer.get(block=True)
+                blue_world_buffer.get(block=True)
```

Comment:
yes so that we're synchronizing with the Simulator binary. We send WorldState (from Thunderscope) and wait to receive a WorldStateReceivedTrigger (sent by the Simulator)